builtin/stash--helper.c

```diff
@@ -634,6 +640,55 @@ static int list_stash(int argc, const char **argv, const char *prefix)
 	return ret;
 }
 
+static int show_stash(int argc, const char **argv, const char *prefix)
+{
+	int i, ret = 0;
+	struct argv_array args = ARGV_ARRAY_INIT;
+	struct argv_array args_refs = ARGV_ARRAY_INIT;
+	struct stash_info info;
+	struct option options[] = {
+		OPT_END()
+	};
+
+	argc = parse_options(argc, argv, prefix, options,
+			git_stash_helper_show_usage, PARSE_OPT_KEEP_UNKNOWN);
+
+	argv_array_push(&args, "diff");
+
+	for (i = 0; i < argc; ++i) {
+		if (argv[i][0] == '-')
+			argv_array_push(&args, argv[i]);
+		else
+			argv_array_push(&args_refs, argv[i]);
+	}
+
+	ret = get_stash_info(&info, args_refs.argc, args_refs.argv);
+	argv_array_clear(&args_refs);
+
+	if (!ret && args.argc == 1) {
+		int showStat = 1, showPatch = 0;
+
+		git_config_get_bool("stash.showStat", &showStat);
+		if (showStat)
+			argv_array_push(&args, "--stat");
+
+		git_config_get_bool("stash.showPatch", &showPatch);
+		if (showPatch)
+			argv_array_push(&args, "-p");
+	}
+
+	if (!ret) {
+		argv_array_pushl(&args, oid_to_hex(&info.b_commit),
+			oid_to_hex(&info.w_commit), NULL);
+
+		free_stash_info(&info);
+
+		ret = cmd_diff(args.argc, args.argv, prefix);
```

Comment:
Just like in the commit before: `cmd_diff()` is not really intended as an API. As such, it might very well not expect to have to clean up open resources or allocated memory, or it might assume that it has to call `setup_git_directory()` itself and simply ignore the `prefix` parameter.

To clarify what that `prefix` and the `run_setup()` is all about:

Some of Git's commands expect to run in a worktree, some of them expect just the presence of a `GIT_DIR` (which can be bare), and some Git commands *might* want to have a worktree or a `GIT_DIR` but might not need any of this. For example, `git diff --no-index <file1> <file2>` works happily outside of any worktree, unattached to any Git repository.

Other commands really *require* a Git repository. They are marked with `RUN_SETUP` in the builtin list in `git.c`. What this does is simply to call `setup_git_directory()`, which historically used `cd("..")` until it found a Git repository, and then returned the relative path to the subdirectory in which it started its search.

For example, if you call `git commit .gitignore` in the subdirectory `Documentation/technical`, `setup_git_directory()` would end up with the top-level directory of the worktree as current directory, and would return `"Documentation/technical/"` as `prefix`, which `cmd_commit()` would then pass on to the code that prefixes `.gitignore` so that you really commit `Documentation/technical/.gitignore`.

Side note: I worked on this recently (I think last year), to allow for `setup_git_directory()` to *not* change the current working directory until the very end (and only if it found the Git repository). But the behavior of `setup_git_directory()` is still the same: discover the Git directory, `cd` to the top-level directory, return the relative path to the previous working directory.

Now, in `cmd_diff()`s case, I think it might actually call `setup_git_directory_gently()`, which will not `die()` if it does not find any Git repository. This is useful e.g. to let `git diff --no-index` use diff settings from the Git config, including the repository config *iff* we found any Git repository.

Back to your use case: I think what we really have to do is to call the diff machinery more directly. Maybe you can figure out from reading the code in `builtin/diff.c` what functions to call? Otherwise, I am fairly certain that there is code in `make_patch()` in `sequencer.c` that should be useful here, too.

Reply:
Thanks for your explanation, @dscho!

In this case, I think there shouldn't be any worries regarding `setup_git_directory()` since using `git stash ...` requires a worktree. (`{ "stash--helper", cmd_stash__helper, RUN_SETUP | NEED_WORK_TREE },`).

The last [commit](https://github.com/git/git/pull/495/commits/0f37f0a6da494a6e1379232d5239eba04d003dc8) I submitted replaces `cmd_diff()` with something more direct.

Reply:
You might indeed get lucky here in that things work as intended. It is still better to imitate `make_patch()`.